### DISCUSSION TITLE
Error in Virtual User 1: can't use non-numeric string as operand of "+"

### DISCUSSION THREAD

**Discussion by j4nd3r53n** [TPC-Council/HammerDB#266] [General]
I have installed HammerDB ver 4.2 on Ubuntu 20.20, and want to run tproc-c on a remote MySQL 5.7 database. I have successfully got it to create the tpcc schema, but when I try to run the workload, it immediately fails with 

`Error in Virtual User 1: can't use non-numeric string as operand of "+"`

I have tried the same against a local MySQL 8.0, and it works without error - what can I do to fix the problem?

**Comment by sm-shaw** [comment]
Hi,
HammerDB will run fine against MySQL 5.7 and 8.0 - there is an example here:
https://www.hammerdb.com/blog/uncategorized/hammerdb-using-mysql-5-7-vs-8-0-to-understand-performance-profiles/
Note that the HammerDB client needs the MySQL 8.0 library, but can then connect fine to MySQL 5.7.
The error "can't use non-numeric string as operand of "+" - suggests that the build didn't complete successfully- 
It is almost certainly because this statement did not return a number as expected
```
 "select max(w_id) from warehouse"
``` 
To troubleshoot log in manually using the username and password you pass to hammerdb and run this statement to see if you get number.

**Reply by j4nd3r53n** [reply]
Thanks for replying! And you are right:

```
mysql> select max(w_id) from warehouse;
+-----------+
| max(w_id) |
+-----------+
|      NULL |
+-----------+
1 row in set (0.01 sec)
```

I'll drop `tpcc` and start over.

**Reply by j4nd3r53n** [reply]
This is a bit strange - It took a long time creating the schema, now it seems to be hanging - the virtual user output shows:

```
Virtual User 1
CREATING TPCC SCHEMA
CREATING DATABASE tpcc
CREATING TPCC TABLES
Loading item
```

and at the bottom of the window:

```
loading history file ... 0 events added
Main console display active (Tcl8.6.10 / tk8.6.10)
The xml is will-formed, applying configuration
(HammerDB-4.2) 1 %
```

`show processlist` in `mysql` shows:

```
| 28 | admin    | 10.1.2.173:8809 | tpcc | Sleep   |    0 | cleaning up | NULL             |
```

This, BTW, is on Aurora serverless running MySQL 5.7; I can't rule out that this may be the problem, but I hope it isn't.

**Reply by j4nd3r53n** [reply]
Further investigation - strace shows that hammerdb in fact write a lot to FD 8, which is the connection to the remote DB - it just doesn't update the log output, it seems. Perhaps this is just s low connection?

**Comment by sm-shaw** [comment]
Hi, yes Aurora is fine. You can also run the build multithreaded so it is the first virtual user that loads ITEM. This table is fairly small so if it is hanging then it could be a network issue as for 1 virtual user this is the first table to load.  Also for AWS it could depend on the instance and the number and type of CPUs available. 

The blog post linked above shows the MySQL 8.0 and 5.7 comparison on a gen -1 architecture. The presentation here https://www.hammerdb.com/about.html shows a MySQL example of 1.7 million NOPM - so this is the sort of upper limit that can be expected for MySQL on an up to date system at present (other databases can offer higher throughput for this workload). 

If the network load is an issue then you can use the datagen feature to generate flat files and create the schema manually.
